crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,398 @@
+//! File containing code for the starcast peer.
+
+use async_trait::async_trait;
+use boringtun::noise::{Tunn, TunnResult};
+use ipnetwork::IpNetwork;
+use std::{net::SocketAddr, time::Duration};
+use telio_crypto::{PublicKey, SecretKey};
+use telio_proto::DataMsg;
+use telio_task::{
+    io::{wait_for_tx, Chan},
+    task_exec, Runtime, RuntimeExt, Task, WaitResponse,
+};
+use telio_utils::interval;
+use telio_utils::{telio_log_error, telio_log_warn};
+use telio_wg::uapi::Peer;
+use tokio::{net::UdpSocket, sync::mpsc::error::SendTimeoutError, time::Interval};
+use x25519_dalek::{PublicKey as PublicKeyDalek, StaticSecret};
+
+/// Constant for maximum packet size.
+const MAX_PACKET: usize = 2048;
+
+#[derive(Debug, thiserror::Error)]
+/// Custom `UdpProxy` error
+pub enum Error {
+    /// `StarcastPeer` Not Configured Error
+    #[error("Starcast peer is not configured")]
+    NotConfigured,
+    /// `StarcastPeer` `IoError`
+    #[error("Starcast peer IoError")]
+    IoError(#[from] std::io::Error),
+    /// Task execution failed
+    #[error(transparent)]
+    Task(#[from] telio_task::ExecError),
+    /// IP network failed
+    #[error("Failed to parse IP network")]
+    IpNetworkError(#[from] ipnetwork::IpNetworkError),
+    /// Send failed
+    #[error(transparent)]
+    Send(#[from] SendTimeoutError<(PublicKey, DataMsg)>),
+    /// Tunnel creation failed
+    #[error("Tunnel creation failed {0}")]
+    TunnNewFail(&'static str),
+}
+
+/// The actual starcast peer containing the task which runs the wait loop for this whole component.
+pub struct StarcastPeer {
+    /// starcast virtual peer task for Telio runtime.
+    task: Task<State>,
+}
+
+/// Starcast virtual peer configuration.
+pub struct Config {
+    /// Public key of the peer with which the starcast virtual peer shall communicate.
+    public_key: PublicKey,
+    /// Port number through which the virtual starcast peer shall communicate with WireGuard.
+    wg_port: u16,
+}
+
+impl StarcastPeer {
+    /// Create and start a new unconfigured starcast virtual peer. Note that it must first be configured to be functional.
+    /// A port is assigned dynamically for it's socket and also a public key which may be obtained by the application
+    /// by using the `get_peer()` method.
+    ///
+    /// # Arguments
+    ///
+    /// * `channel` - Channel for communicating decapsulated packets to and from the multicaster.
+    ///
+    /// # Returns
+    ///
+    /// A new starcast virtual peer.
+    pub async fn start(channel: Chan<Vec<u8>>, allowed_ips: Vec<IpNetwork>) -> Result<Self, Error> {
+        // Port 0 means the OS will dynamically allocate port.
+        const TIMER_UPDATE_PERIOD: Duration = Duration::from_millis(25);
```

Comment:
Will waking up so often, not kill any battery updates? Can we replace it with some sort of notification mechanism or at least significantly increase the period?

Reply:
It seems that boringtun updates them every 250ms so probably we can do the same. Inside update_timers they anyway only care about seconds passing so 25 is overkill for sure. Comment from the code: 
`
// All the times are counted from tunnel initiation, for efficiency our timers are rounded to a second, as there is no real benefit to having highly accurate timers
`

Reply:
I also saw that comment, but wasn't sure how to interpret it in our situation. So I increased the interval to 250 ms for now.

Reply:
Is there no way to be notified instead of polling?

Reply:
The .update_timers() function should be called at least as often as the keep alive packets need to be handled, so I thought it'd be best to call it periodically, tho I'm open to ideas.

Reply:
I don't see a way to subscribe for the keep alive packets without also waking up and calling `update_timers`. In that case I think that 1s interval should still be fine but if you prefer to keep it at 0.25 it's also fine by me.